File: 算法/代码随想录算法PDF/00调试源代码/77.组合.py
#
# @lc app=leetcode.cn id=77 lang=python3
#
# [77] 组合
#

# @lc code=start
class Solution:
    def combine(self, n: int, k: int) -> List[List[int]]:
        path = []
        res = []
        def backtracking(startIndex):
            nonlocal n, k, path, res
            if len(path) == k:
                res.append(path[:])
                return
            for i in range(startIndex, n + 1):
                if i > n - (k - len(path)) + 1:  # 剪枝
                    break
                path.append(i)
                backtracking(i + 1)
                path.pop()
        
        backtracking(1)
        return res
# @lc code=end

File: 算法/代码随想录算法PDF/00调试源代码/77.组合.cpp
/*
 * @lc app=leetcode.cn id=77 lang=cpp
 *
 * [77] 组合
 */

// @lc code=start
class Solution {
public:
    vector<int> path;
    vector<vector<int>> res;
    int globalN, globalK;
    void backtracking(int startIndex) {
        if (path.size() == globalK) {
            res.push_back(path);
            return ;
        }
        for (int i = startIndex; i <= globalN - (globalK - path.size()) + 1; i++) {
            path.push_back(i);
            backtracking(i + 1);
            path.pop_back();
        }
    }
    vector<vector<int>> combine(int n, int k) {
        globalN = n;
        globalK = k;
        backtracking(1);
        return res;
    }
};
// @lc code=end

